File: 백준/Silver/2531. 회전 초밥/회전 초밥.py
import sys
input=sys.stdin.readline

# k개 접시 연속해 먹으면 할인가격
# k개 접시 연속해 먹으면 쿠폰 초밥 무료 제공(없다면 만들어 제공)

def update(item):
  if item in info:
    info[item]+=1
  else:
    info[item]=1

N,d,k,c=map(int,input().split())
l=[]
for i in range(N):
  l.append(int(input()))
l=l+l

info={}
for i in range(k):
  update(l[i])

answer=len(info.keys())
if c not in info:
  answer+=1

for i in range(1,N):
  update(l[i+k-1])
  if info[l[i-1]]==1:
    info.pop(l[i-1])
  else:
    info[l[i-1]]-=1
  now=len(info.keys())
  if c not in info:
    now+=1
  answer=max(answer,now)

print(answer)

File: 백준/Gold/15961. 회전 초밥/회전 초밥.js
const readline=require('readline')
const rl=readline.createInterface({
  input:process.stdin,
  output:process.stdout
})

const input=[]
let iter=0
rl.on('line',(line)=>{
  input.push(line)
}).on('close',()=>{
  let [N,d,k,c]=input[iter++].split(' ').map(item=>Number(item))
  let l=[]
  for(let i=0;i<N;i++){
    l.push(Number(input[iter++]))
  }
  l=[...l,...l]

  const info={[c]:1}
  for(let i=0;i<k;i++){
    if (l[i] in info) info[l[i]]+=1
    else info[l[i]]=1
  }
  let answer=Object.keys(info).length
  let cnt=answer

  for(let i=1;i<N;i++){
    const left=l[i-1]
    info[left]-=1
    if (info[left]===0) {
      delete(info[left])
      cnt-=1
    }
    const value=l[i+k-1]
    if (value in info) info[value]+=1
    else {
      info[value]=1
      cnt+=1
    }
    if (answer<cnt) answer=cnt
  }

  console.log(answer)
  process.exit()
})
